### ISSUE TITLE
Investigating issue with running cascading Annotation Processors on generated Immutables in Eclipse

### ISSUE THREAD

**Issue by ChrisTrenkamp** [immutables/immutables#1445]
https://bugs.eclipse.org/bugs/show_bug.cgi?id=527420

I have a situation that exactly describes the above issue.  I'm injecting some Annotations on our *Impl classes via `@InjectAnnotation`, and those injected Annotations are further processed with our own Annotation Processor.  Like how the above issue describes, Eclipse will generate all the java files, but it will not compile some of them, causing missing class files and compilation errors.  

If I go to each Immutable definition that's missing its *Builder/*Impl class file and force Eclipse to recompile it by making an insignificant edit, that will fix the compilation errors.  However, when we do a Clean Build or Gradle Refresh, the class files will go missing again.

This issue does not occur when we compile with Gradle on the commandline.  

I'm trying to figure out what's going on here.  In the above issue, one of the Eclipse developers said that they have plenty of test cases for cascading Annotation Processors.  I'll assume that Eclipse is behaving properly.  Can anyone help pin down what the issue might be?

Here's some of the notes I've gathered while investigating this:

* Does Immutables directly generate .class files?  

One of the dev's in the above issue asked this, and commented that, if so, that part of the spec is ill-defined and not very well tested in Eclipse.

* Does Immutables generate any files after `roundEnv.processingOver()`?  

https://code.google.com/archive/p/acris/wikis/CodeGenerationPlatform_Pitfall_Rounds.wiki

This person had issues getting Eclipse to run his annotation processor, and it was fixed after he respected the `processingOver` flag.  I see that Immutables checks for this flag and runs some post-processing operations:

https://github.com/immutables/immutables/blob/master/generator/src/org/immutables/generator/AbstractGenerator.java#L97
https://github.com/immutables/immutables/blob/master/generator/src/org/immutables/generator/EnvironmentState.java#L85

I can't tell what's actually going on in the `completeProcessing` method.  Does it generate any additional files in that method?

**Comment by elucash** [comment]
> Does Immutables directly generate .class files?
no 
> Does Immutables generate any files after roundEnv.processingOver()?
let's see

```java
// in AbstractGenerator
public final boolean process(Set<? extends TypeElement> annotations, RoundEnvironment round) {
    try {
      StaticEnvironment.init(annotations, round, processingEnv);
      if (!round.processingOver() && !round.errorRaised()) {
        process();
      }
      StaticEnvironment.shutdown();
    } catch (Exception ex) {
      processingEnv.getMessager()
          .printMessage(Diagnostic.Kind.ERROR,
              getClass().getName() + " threw " + Throwables.getStackTraceAsString(ex));
    }
    return false;
  }
```
where `shutdown()` completes some resources
Apparently, `META-INF/services/` are generated on `shutdown()`, but 
java files are generated inline, during processing, not after

```java
 // from org.immutables.generator.Output
  public final Templates.Invokable java = new Templates.Invokable() {
    @Override
    public Invokable invoke(Invokation invokation, Object... parameters) {
      String packageName = parameters[0].toString();
      String simpleName = parameters[1].toString();
      Element originatingElement = (Element) parameters[2];
      Invokable body = (Invokable) parameters[3];

      ResourceKey key = new ResourceKey(packageName, simpleName, Delegates.unwrap(originatingElement));
      SourceFile javaFile = getFiles().sourceFiles.get(key);
      body.invoke(new Invokation(javaFile.consumer));
      javaFile.complete();     // <-------   here
      return null;
    }
  };
```

So it seems that we generate java files inline, during rounds. But I suggest to debug it to somehow make sure it does. What about Javac ? same or different from ECJ. Also, have you tried batch/standalone ECJ jar to compile?

**Comment by ChrisTrenkamp** [comment]
This issue continues to remain a mystery to me.  I wasn't able to any of the following:

* Build the immutables library from the commandline 
* Attach a debugger to the build process
* Build with ECJ

... but this issue seems to be fixed after I made some changes to the code:

* I discovered my annotation processor was potentially generating the same class twice.  Perhaps that's what caused Eclipse's annotation processor implementation to stumble over itself.
* My code was originally accessing the generated code by statically initializing it:
```java
  private static final MyGeneratedClass SOME_GENERATED_CLASS = new MyGeneratedClass();

...
    return SOME_GENERATED_CLASS .doSomething(...);
...
```

I changed the generated class to be a self-contained singleton:
```java
...
  return MyGeneratedClass.doSomething(...);
...
```

I doubt the second case had anything to do with it.

So if anyone else runs into the same issue, double-check your annotation processor and make sure it doesn't generate the same class twice.
